### PR TITLE
HDDS-5836. [Multi-Tenant] Use Ranger java client.

### PR DESCRIPTION
**Pull request by aswinshakil**
## What changes were proposed in this pull request?

- Add implementation of `MultiTenantAccessController` using the [Ranger client](https://github.com/apache/ranger/tree/master/intg/src/main/java) instead of HTTP REST calls.
- Remove the `MultiTenantAccessController` REST implementation for now. If the Ranger client turns out not to be ready yet we can finish the REST implementation and continue using it.
- Add an `InMemoryMultiTenantAccessController` for testing that simulates Ranger by storing state in memory.
- Add/modify operations in `MultiTenantAccessController` to support only our current needs from Ranger.

## What is the link to the Apache JIRA

HDDS-5836

## How was this patch tested?

- [x]  Test of interface and builders added using the in memory controller.
- [x]  Running the tests in a live secure cluster using the Ranger client controller.

### CODE REVIEW

**Comment by errose28** [comment]
Converting to draft while we debug some Recon dependency errors and update Jar report.

File: hadoop-ozone/dist/src/main/license/jar-report.txt [hunk 1]
```diff
@@ -25,33 +29,43 @@ share/ozone/lib/commons-configuration2.jar
...
+share/ozone/lib/grpc-protobuf.jar
 share/ozone/lib/grpc-stub.jar
 share/ozone/lib/gson.jar
 share/ozone/lib/guava-jre.jar
 share/ozone/lib/guice-assistedinject.jar
 share/ozone/lib/guice-bridge.jar
-share/ozone/lib/guice.jar
 share/ozone/lib/guice-multibindings.jar
 share/ozone/lib/guice-servlet.jar
+share/ozone/lib/guice.jar
```
Comment (smengcl) [inline hadoop-ozone/dist/src/main/license/jar-report.txt:68]: nit: can we avoid re-ordering the lines that indicates the same jar and keep only the lines added (or removed if any)

File: hadoop-ozone/ozone-manager/pom.xml [hunk 2]
```diff
@@ -144,6 +144,24 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
       <version>0.9.11</version>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.ranger</groupId>
+      <artifactId>ranger-intg</artifactId>
+      <version>3.0.0-SNAPSHOT</version>
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/pom.xml:150]: IIRC the feature we need will also be released in Ranger 2.3.0. If there is a 2.3.0-SNAPSHOT jar, can we use that (and try it) instead?

File: hadoop-ozone/ozone-manager/pom.xml [hunk 3]
```diff
@@ -144,6 +144,24 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
...
+      <exclusions>
+        <exclusion>
+          <groupId>ch.qos.logback</groupId>
+          <artifactId>logback-classic</artifactId>
+        </exclusion>
+        <exclusion>
+          <groupId>com.sun.jersey</groupId>
+          <artifactId>jersey-bundle</artifactId>
+        </exclusion>
+      </exclusions>
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/pom.xml:163]: Are `elasticsearch` and `lucene` required, which seems to be introduced as transitive dependencies? If not we could exclude them as well?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/RangerClientMultiTenantAccessController.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+  private final Map<String, IAccessAuthorizer.ACLType> stringToAcl;
+  private final String omPrincipal;
+
+  public RangerClientMultiTenantAccessController(OzoneConfiguration conf)
+      throws IOException {
+    aclToString = MultiTenantAccessController.getRangerAclStrings();
+    stringToAcl = new HashMap<>();
+    aclToString.forEach((type, string) -> stringToAcl.put(string, type));
+
+    // TODO: Double check configuration set up after HDDS-6612 is merged.
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/RangerClientMultiTenantAccessController.java:68]: Now that HDDS-6612 is merged. Would you merge the feature branch into this PR and then revisit this comment?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/RangerClientMultiTenantAccessController.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    rangerServiceName = conf.get(OZONE_RANGER_SERVICE);
+    String configuredOmPrincipal = conf.get(OZONE_OM_KERBEROS_PRINCIPAL_KEY);
+    // Replace _HOST pattern with host name in the Kerberos principal. Ranger
+    // client currently does not do this automatically.
+    omPrincipal = SecurityUtil.getServerPrincipal(
+        configuredOmPrincipal, OmUtils.getOmAddress(conf).getHostName());
+    String keytabPath = conf.get(OZONE_OM_KERBEROS_KEYTAB_FILE_KEY);
+
+    client = new RangerClient(rangerHttpsAddress,
+        "kerberos", omPrincipal, keytabPath, rangerServiceName, "ozone");
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/RangerClientMultiTenantAccessController.java:79]: nit

```suggestion
    client = new RangerClient(rangerHttpsAddress, KERBEROS.name().toLowerCase(),
        omPrincipal, keytabPath, rangerServiceName, OzoneConsts.OZONE);
```

(don't forget the imports)

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/multitenant/InMemoryMultiTenantAccessController.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.Optional;
+
+/**
+ * In Memory version of MultiTenantAccessController.
+ */
+public class InMemoryMultiTenantAccessController
+    implements MultiTenantAccessController {
+
+  private final Map<String, Policy>  policies;
+  private final Map<String, Role>  roles;
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/multitenant/InMemoryMultiTenantAccessController.java:33]: nit
```suggestion
  private final Map<String, Policy> policies;
  private final Map<String, Role> roles;
```

File: hadoop-ozone/dist/src/main/license/jar-report.txt (hunk 1 above)
Comment (errose28) [reply hadoop-ozone/dist/src/main/license/jar-report.txt:68]: This file is updated automatically by running `update-jar-report.sh`. I believe the order changes due to system specific implementations of `sort`, which determines things like whether of not `share/ozone/lib/guice.jar` comes before or after `share/ozone/lib/guice-multibindings.jar`.

File: hadoop-ozone/ozone-manager/pom.xml (hunk 2 above)
Comment (errose28) [reply hadoop-ozone/ozone-manager/pom.xml:150]: I see that version on Ranger's ranger-2.3 release branch but it looks like it is not being published
```
[ERROR] Failed to execute goal on project ozone-manager: Could not resolve dependencies for project org.apache.ozone:ozone-manager:jar:1.3.0-SNAPSHOT: Could not find artifact org.apache.ranger:ranger-intg:jar:2.3.0-SNAPSHOT in apache.snapshots.https (https://repository.apache.org/content/repositories/snapshots) -> [Help 1]
```

FYI the Ranger 2.3.0 release is being tracked in this Jira: https://issues.apache.org/jira/browse/RANGER-3710

Comment (smengcl) [reply hadoop-ozone/ozone-manager/pom.xml:150]: Sounds good to me. We can keep 3.0.0-SNAPSHOT for now.

**Review by smengcl — APPROVED** [review]
+1 pending CI

**Comment by smengcl** [comment]
Thanks @errose28 and @aswinshakil for working on this!

**Pull request merged**